iroha/src/kura.rs

```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
+                let new_block = last_block.header().height;
+                if prev_block != new_block + 1 {
+                    return Err(eyre::eyre!("Incosequential blocks were found: [{} - {}]. Data storage is probably corrupt.", prev_block, new_block));
+                }
+            } else if prev_block != 1_u64 {
+                return Err(eyre::eyre!(
+                    "First block not found. Data storage is probably corrupt."
+                ));
+            }
+            blocks.append(&mut read_blocks);
+        }
+        iroha_logger::info!("Read {} blocks from block store.", blocks.len());
+        Ok(blocks)
     }
+}
 
-    /// Returns a sorted vector of blocks starting from 0 height to the top block.
-    #[iroha_futures::telemetry_future]
-    pub async fn read_all(&self) -> Vec<VersionedCommittedBlock> {
-        let mut height = 1;
-        let mut blocks = Vec::new();
-        while let Ok(block) = self.read(height).await {
-            blocks.push(block);
-            height += 1;
+/// Returns deserialized blocks from specific file
+async fn read_all_from_storage_file(path: &Path) -> Result<Vec<VersionedCommittedBlock>> {
+    let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+    let mut file = fs::File::open(&path).await.wrap_err("No file found.")?;
+    let metadata = fs::metadata(&path)
+        .await
+        .wrap_err("Unable to read metadata.")?;
+    #[allow(clippy::cast_possible_truncation)]
+    let mut buffer = vec![0; metadata.len() as usize];
+    let _ = file.read(&mut buffer).await.wrap_err("Buffer overflow.")?;
+    while let Some((msg, tail)) = pop_from_buffer(&mut buffer)? {
+        let block = VersionedCommittedBlock::decode_versioned(msg)
+            .wrap_err("Failed to read block from store.")?;
+
+        if let Some(last_block) = blocks.last() {
+            let prev_block = last_block.header().height;
+            let new_block = block.header().height;
+            if new_block != prev_block + 1 {
+                return Err(eyre::eyre!("Incosequential blocks were found: [{} - {}]. Data storage is probably corrupt.", prev_block, new_block));
+            }
         }
-        iroha_logger::info!("Read {} blocks from block store.", blocks.len());
-        blocks
+        blocks.push(block);
+        buffer = tail.to_vec();
+    }
+    Ok(blocks)
+}
+
+fn pop_from_buffer(mut buffer: &mut [u8]) -> Result<Option<(&[u8], &[u8])>> {
+    if buffer.is_empty() {
+        return Ok(None);
     }
+
+    let (int_bytes, rest) = buffer.split_at_mut(4);
+    buffer = rest;
+    let block_size = u32::from_le_bytes((*int_bytes).try_into()?);
+
+    let (msg, tail) = buffer.split_at_mut(block_size as usize);
+    Ok(Some((msg, tail)))
+}
+
+/// Returns sorted Vec of datafiles base heights as u64
+///
+/// # Errors
+/// Will fail on filesystem access error
+///
+async fn storage_files_base_indices(path: &Path) -> Result<BTreeSet<NonZeroU64>> {
```

Comment:
For some reason it broke :(
I don't see a reason why - my best guess is something in uses, but I could be wrong.